File: python/tracing/together/01_chat_completion.py
from __future__ import annotations

from _shared import (
    example_attributes,
    first_message_text,
    make_client,
    make_custom_identifier,
    make_respan,
    model_name,
    print_result,
    print_start,
    workflow_name,
)
from respan import workflow

EXAMPLE_NAME = "chat-completion"


@workflow(name=workflow_name(EXAMPLE_NAME))
def _chat_completion_workflow(prompt: str) -> str:
    with make_client() as client:
        response = client.chat.completions.create(
            model=model_name(),
            messages=[{"role": "user", "content": prompt}],
            max_tokens=80,
            temperature=0,
        )
        return first_message_text(response)


def run_chat_completion() -> None:
    custom_identifier = make_custom_identifier(EXAMPLE_NAME)
    respan = make_respan(EXAMPLE_NAME, custom_identifier)
    text = ""

    try:
        with example_attributes(EXAMPLE_NAME, custom_identifier):
            print_start(EXAMPLE_NAME, custom_identifier)
            text = _chat_completion_workflow(
                "Reply with one concise sentence about tracing Together AI apps."
            )
    finally:
        respan.shutdown()

    print_result(EXAMPLE_NAME, custom_identifier, text)


if __name__ == "__main__":
    run_chat_completion()

File: typescript/tracing/together-ai/01_chat_completion.ts
import {
  MODELS,
  captureFeature,
  createRespan,
  createTogether,
  logExampleResult,
  runWithTogetherWorkflow,
  shutdownRespan,
  summarizeChatCompletion,
  withTogetherRetries,
} from "./_shared.js";

const workflowName = "together-ai-ts-chat-completion";
const respan = createRespan("together-ai-typescript-chat");

try {
  const details = await runWithTogetherWorkflow(respan, workflowName, async () => {
    const client = createTogether();
    return await captureFeature(workflowName, async () => {
      const response = await withTogetherRetries(workflowName, async () =>
        await client.chat.completions.create({
          model: MODELS.chat,
          messages: [
            { role: "system", content: "Answer in one concise sentence." },
            { role: "user", content: "Say that Respan Together AI chat tracing succeeded." },
          ],
          max_tokens: 64,
          temperature: 0,
        }),
      );
      return summarizeChatCompletion(response);
    });
  });

  logExampleResult(workflowName, details);
} finally {
  await shutdownRespan(respan);
}
